File: flutter_readium/ios/flutter_readium/Sources/flutter_readium/model/GuidedNavigationText.swift
import Foundation

/// Text associated with a guided navigation object or description.
///
/// Can be either a plain string or a structured object with optional SSML
/// markup and BCP 47 language tag.
///
/// See https://readium.org/guided-navigation/schema/text.schema.json
enum GuidedNavigationText: Equatable {
  case string(String)
  case object(plain: String?, ssml: String?, language: String?)

  /// Parses a [GuidedNavigationText] from its JSON representation.
  ///
  /// Accepts either a String or a dictionary with at least one of `plain` or `ssml`.
  static func fromJson(_ json: Any?) -> GuidedNavigationText? {
    if let s = json as? String {
      return s.isEmpty ? nil : .string(s)
    }
    if let obj = json as? [String: Any] {
      let plain = obj["plain"] as? String
      let ssml = obj["ssml"] as? String
      let language = obj["language"] as? String
      guard plain != nil || ssml != nil else { return nil }
      return .object(plain: plain, ssml: ssml, language: language)
    }
    return nil
  }

  func toJson() -> Any {
    switch self {
    case .string(let s):
      return s
    case .object(let plain, let ssml, let language):
      var res: [String: Any] = [:]
      if let plain = plain { res["plain"] = plain }
      if let ssml = ssml { res["ssml"] = ssml }
      if let language = language { res["language"] = language }
      return res
    }
  }
}

File: flutter_readium_platform_interface/lib/src/shared/guided_navigation/guided_navigation_text.dart
import 'package:equatable/equatable.dart';
import 'package:meta/meta.dart';

import '../../utils/jsonable.dart';

/// Text associated with a guided navigation object or description.
///
/// Can be either a plain string or a structured object with optional SSML
/// markup and language tag.
///
/// See https://readium.org/guided-navigation/schema/text.schema.json
sealed class GuidedNavigationText implements JSONable {
  const GuidedNavigationText();

  /// Parses a [GuidedNavigationText] from its JSON representation.
  ///
  /// Accepts either a [String] or a [Map<String, dynamic>] with at least one
  /// of [plain] or [ssml].
  static GuidedNavigationText? fromJson(dynamic json) {
    if (json == null) return null;
    if (json is String) {
      return json.isNotEmpty ? GuidedNavigationTextString(json) : null;
    }
    if (json is Map<String, dynamic>) {
      final jsonObject = Map<String, dynamic>.of(json);
      final plain = jsonObject.optNullableString('plain', remove: true);
      final ssml = jsonObject.optNullableString('ssml', remove: true);
      final language = jsonObject.optNullableString('language', remove: true);
      if (plain == null && ssml == null) return null;
      return GuidedNavigationTextObject(
        plain: plain,
        ssml: ssml,
        language: language,
      );
    }
    return null;
  }
}

/// A guided navigation text represented as a plain string.
@immutable
final class GuidedNavigationTextString extends GuidedNavigationText with EquatableMixin {
  const GuidedNavigationTextString(this.value);

  final String value;

  @override
  List<Object> get props => [value];

  @override
  String toJson() => value;
}

/// A guided navigation text represented as a structured object with optional
/// SSML markup and BCP 47 language tag.
@immutable
final class GuidedNavigationTextObject extends GuidedNavigationText with EquatableMixin {
  const GuidedNavigationTextObject({this.plain, this.ssml, this.language});

  final String? plain;
  final String? ssml;
  final String? language;

  @override
  List<Object?> get props => [plain, ssml, language];

  @override
  Map<String, dynamic> toJson() => {}
    ..putOpt('plain', plain)
    ..putOpt('ssml', ssml)
    ..putOpt('language', language);
}
